Fix KCFI types for generated functions with integer normalization

samitolvanen opened the pull request:
With -fsanitize-cfi-icall-experimental-normalize-integers, Clang appends ".normalized" to KCFI types in CodeGenModule::CreateKCFITypeId, which changes type hashes also for functions that don't have integer types in their signatures. However, llvm::setKCFIType does not take integer normalization into account, which means LLVM generated functions with KCFI types, e.g. sanitizer constructors, will fail KCFI checks when integer normalization is enabled in Clang.

Add a kcfi-normalized module flag to indicate integer normalization is used, and append ".normalized" to KCFI types also in llvm::setKCFIType to fix the type mismatch.

cc @rcvalle

samitolvanen committed:
Fix KCFI types for generated functions with integer normalization

With -fsanitize-cfi-icall-experimental-normalize-integers, Clang
appends ".normalized" to KCFI types in CodeGenModule::CreateKCFITypeId,
which changes type hashes also for functions that don't have integer
types in their signatures. However, llvm::setKCFIType does not take
integer normalization into account, which means LLVM generated
functions with KCFI types, e.g. sanitizer constructors, will fail KCFI
checks when integer normalization is enabled in Clang.

Add a kcfi-normalized module flag to indicate integer normalization is
used, and append ".normalized" to KCFI types also in llvm::setKCFIType
to fix the type mismatch.

rcvalle commented:
Thank you for your time and for working on this, @samitolvanen! Much appreciated.

How about changing the module flag name to "cfi-normalize-integers" to be more consistent with the naming convention used, and as these could've also possibly happened in regular CFI and with user space programs too (and also possibly with generalize pointers, but we can take a look at it later/separately)?

MaskRay commented:
>  However, llvm::setKCFIType does not take integer normalization into account, which means LLVM generated functions with KCFI types, e.g. sanitizer constructors, will fail KCFI checks when integer normalization is enabled in Clang [...]

LG. It's best to have a test for the LLVM generated function, not just reuse kcfi-normalize.c whose module flags could be puzzling for readers.

samitolvanen committed:
amend! Fix KCFI types for generated functions with integer normalization

Fix KCFI types for generated functions with integer normalization

With -fsanitize-cfi-icall-experimental-normalize-integers, Clang
appends ".normalized" to KCFI types in CodeGenModule::CreateKCFITypeId,
which changes type hashes also for functions that don't have integer
types in their signatures. However, llvm::setKCFIType does not take
integer normalization into account, which means LLVM generated
functions with KCFI types, e.g. sanitizer constructors, will fail KCFI
checks when integer normalization is enabled in Clang.

Add a cfi-normalize-integers module flag to indicate integer
normalization is used, and append ".normalized" to KCFI types also in
llvm::setKCFIType to fix the type mismatch.

samitolvanen committed:
fixup! Fix KCFI types for generated functions with integer normalization

samitolvanen commented:
> How about changing the module flag name to "cfi-normalize-integers" to be more consistent with the naming convention used

Sure, renamed.

samitolvanen commented:
> LG. It's best to have a test for the LLVM generated function, not just reuse kcfi-normalize.c whose module flags could be puzzling for readers.

I think we still need to test that Clang adds the module flag, but I added another test for the GCOV functions to make sure the hash is expected with and without integer normalization. PTAL.

rcvalle commented:
LGTM. Thanks again for your time and for working on this, @samitolvanen! Much appreciated.

samitolvanen commented:
/cherry-pick e1c36bde0551977d4b2efae032af6dfc4b2b3936

samitolvanen commented:
It would be nice to get this cherry-picked to 19.x, so the Rust kernel folks don't have to wait for LLVM 20: https://lore.kernel.org/lkml/202410281414.c351044e-oliver.sang@intel.com/

samitolvanen commented:
/cherry-pick e1c36bde0551977d4b2efae032af6dfc4b2b3936

Darksonn commented:
This [has landed in 19.1.3](https://discourse.llvm.org/t/llvm-19-1-3-relased/82829). Thanks!
